Source/Editor/DualityEditor/PackageManagement/PackageManager.cs

```diff
@@ -232,29 +225,26 @@ public void VerifyPackage(LocalPackage localPackage)
 		/// </summary>
 		public void UninstallNonRegisteredPackages()
 		{
-			IPackage[] allInstalledPackages = this.manager
-				.LocalRepository
-				.GetPackages()
-				.ToArray();
+			IEnumerable<PackageIdentity> allInstalledPackages = this.manager.GetInstalledPackages();
```

Comment:
GetInstalledPackages relies on reading out the package.config file to see what packages are installed however duality also has logic for this already and iam not sure how to merge these in a clean way. @ilexp

Reply:
So I cannot simply remove that logic from the manager as that would break this code

Reply:
Ah I think that a `installed package` does not mean that the package is registered in the package.config but that its actually restored to the source/packages folder.

However with the new nuget this doesn't happen anymore and thus clashes a bit with this approach. Ofcourse one can still add this back in manually...

Reply:
So I fixed the above problem by looking at what packages are actually restored but now I get this
![image](https://user-images.githubusercontent.com/19387223/62644443-65507180-b94a-11e9-86ad-f82740f7f7c5.png)

Reply:
Ugh yet again duality is removing packages that shouldn't be removed. That package that the error is complaining about is actually installed but then later on removed by duality.

@ilexp idk how to fix this, why is duality modifying the packages folder in the first place? That is nugets reponsibility and not duality. I feel duality is doing way too much and that is now all getting in the way of this upgrade.

Reply:
@ilexp I think we need to redesign the packagemanager logic around these points:
- Nuget manages the packages folder (or global packages folder in v5.x)
- Nuget manages the packages.config
- Duality manages the plugin and lib folders

From a high level this should result in the following separation
1. A NugetPackageManager class that provides all the basic nuget operations (install, uninstall, upgrade, restore etc). This class protects the rest of duality from any changes on the nuget side (such as another upgrade).
2. A class that provides duality specific logic such as copying packages to the plugin/lib folders. If this class needs to do something with nuget packages or the packages.config (what packages are installed?) it will ask NugetPackageManager for this.
3. Other utility/glue/etc classes

Reply:
Only skimmed the above thread since I'm very low on time right now, but it seems like you somehow got off the wrong track somewhere. Short version from what I gathered:

> Nuget manages the packages folder (or global packages folder in v5.x)

That's already the case.

> Nuget manages the packages.config

That's already the case as well.

> Duality manages the plugin and lib folders

And this one too!

The only way in which you could say that Duality handles the first two points is via the actual NuGet API, which is intended for that sort of thing. It's possible that this API no longer exists, or some of the concepts changed in newer NuGet, but it's not intention that Duality would do either of those - and as far as I remember, it shouldn't in the current implementation.

The outline of the current / old implementation that is probably related to some of the issues that arose:

- Duality has a list of (Duality-only!) packages in `PackageConfig.xml`. This is a letter of intent which (Duality only!) packages there should be.
- Duality sets up NuGet with the local `Packages` folder and asks it which packages are currently installed. Internally, I assume NuGet figures that out by checking folders and files.
- Duality then asks Nuget to remove those that aren't on the list, and to install those that are, in order to sync whatever state the NuGet stuff is in with the intended state that the Duality config specifies. Dependencies and everything are handled by NuGet, Duality has no knowledge of packages that wouldn't show up in its own package manager.
- To apply the package install to plugin and lib folders, Duality registers a NuGet event that fires when packages are installed or uninstalled, and uses the provided info to copy files as needed.

In the old NuGet at least, "install" meant the package is physically in the right place and extracted, not in some config list, and "uninstall" meant that it is physically removed from there.

Let me know if that helps, will try to check back soon-ish.

Reply:
> Duality then asks Nuget to remove those that aren't on the list, and to install those that are, in order to sync whatever state the NuGet stuff is in with the intended state that the Duality config specifies. Dependencies and everything are handled by NuGet, Duality has no knowledge of packages that wouldn't show up in its own package manager.

So this is causing it to break atm since its removing packages that duality packages depend on that shouldnt be removed. I don't know how to fix this.

Reply:
Hmm, it [shouldn't remove any non-Duality packages](https://github.com/AdamsLair/duality/blob/debb16086e6a07e86ea622c392eba3a1de47753a/Source/Editor/DualityEditor/PackageManagement/PackageManager.cs#L244-L245), and even if it removes a Duality package that later turns out to be a dependency of one that _is_ needed, that [should be fixed right after](https://github.com/AdamsLair/duality/blob/debb16086e6a07e86ea622c392eba3a1de47753a/Source/Editor/DualityEditor/Program.cs#L145-L155) when it verifies that every expected package is installed. It's more of an edge case though that doesn't happen that often, so I'd guess it's some sort of a bug that the package is removed in the first place.

As soon as I manage to carve out some time, I'll have a look as well, see if I can help.

Reply:
hmm maybe I broke something in that area then

Reply:
> What about just ignore cleaning the package folder btw?

The reason it's doing that is not just cleaning the package folder, but a necessary step (for the current implementation anyway) to find out which actual files to remove from plugin and lib folders via the events that NuGet 2.x fires on uninstall, including dependencies. 

For the current implementation, the packages folder isn't a side effect or cache we could just ignore, but NuGet's "working memory". From what I gather from your progress reports and comments, this is no longer the case anyway, so I assume some things on the implementation will change in that regard. What we still need is some automated uninstall of Duality packages (+ deps) that are no longer in the `PackageConfig.xml` despite being previously installed - and vice versa, an auto-install for those that are in there, but not installed yet.

Use cases:

- Pull update from version control, run editor. It should match the local packages with what's checked in via `PackageConfig.xml`, i.e. account for added and removed ones.
- User manually edits `PackageConfig.xml`, then starts the editor.
- Something in the package manager goes really wrong and Duality crashes. A restart can bring packages back in order.

So, assuming the new NuGet doesn't have any equivalent to a local package folder / working memory, then the Duality package manager will need some way to find out

- Which files to copy or remove when a package is installed / uninstalled. Without having looked at the code, I assume you got that covered for a while via NuGet API.
- Which items in `PackageConfig.xml` are not currently present in the project and need an install.
- Which packages that are currently present in the project are not in `PackageConfig.xml` and need an uninstall.

In other words, there needs to be a way to ask for the "status quo" of current package installs, so we can diff it with the "letter of intent" in the `PackageConfig.xml` to find out what needs to be done. Previously, that was achieved via NuGet API (`this.manager.LocalRepository`, which used the packages folder) and then encapsulated in Duality [local setup](https://github.com/AdamsLair/duality/blob/debb16086e6a07e86ea622c392eba3a1de47753a/Source/Editor/DualityEditor/PackageManagement/PackageManager.cs#L667-L682) data. Is there any equivalent in NuGet 5, which we can query for project "status quo"? 

If there isn't, we could come up with something custom, like a second, internal package config representing the actual local setup, or similar.